### DISCUSSION TITLE
git-describe --tags reporting incorrect offset?

### DISCUSSION THREAD

**Discussion by pete4abw** [community/community#39358] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

Within my LOCAL repo, I merged between two branches and then did `git reset --hard` to revert the merge back to the pre-merge HEAD. When I accepted the fast-forward merge , HEAD had a message "Merge branch...". When I ran  
`git describe --tags`, it showed an offset of 3 from the tag.

I assigned the tag v0.9.3 in the source branch and then performed the merge. It does show in the destination branch `lzma-22.01`, but seems to reference the tag in the source branch.

```
$ git log -n5 HEAD
 3784ebe 2022-11-15  (HEAD -> lzma-22.01, origin/lzma-22.01) Merge branch 'whats-next' into lzma-22.01
 38294ae 2022-11-15  (tag: v0.9.3, origin/whats-next, whats-next) VERSION Bump 0.9.3. Doc Update.
 f7a5891 2022-11-15  Remove direct fprintf. Use print macros in lrzip_private.h only.
 24c759c 2022-10-17  Merge branch 'whats-next' into lzma-22.01
 733c93f 2022-10-17  FIX: Keep writing output even when input > 2GB
$ git log -n5 whats-next
 38294ae 2022-11-15  (tag: v0.9.3, origin/whats-next, whats-next) VERSION Bump 0.9.3. Doc Update.
 f7a5891 2022-11-15  Remove direct fprintf. Use print macros in lrzip_private.h only.
 733c93f 2022-10-17  FIX: Keep writing output even when input > 2GB
 9fb7258 2022-08-23  (tag: v0.9.2) VERSION Bump 0.9.2. Doc Update.
 69c806d 2022-08-18  Temporary directory now defaults to /tmp.
$ git describe --tags
v0.9.3-3-g3784ebe
```

There are four questions here.
1. How to merge without the merge message?
2. Why is the offset from the tag 3?
3. Can I correct? If so, how?
4. What did I do wrong?

Thank you.

**Comment by airtower-luna** [comment, marked as answer]
> How to merge without the merge message?

You need to fast-forward merge. This is the default, but works only if the HEAD of the current branch is a *direct ancestor* of the branch you're merging. Your `lzma-22.01` branch had `24c759c`, which wasn't on `whats-next`, so it wasn't possible. You can use `git merge --ff-only` to make Git either do a fast-forward merge or fail.

> Why is the offset from the tag 3?

It's the number of commits on your current branch that aren't part of the tagged history. The log snippets you posted show two merge commits on `lzma-22.01`, I guess there's a third somewhere in the older history.

> Can I correct? If so, how?

You need to bring the branches to the exact same commit, and then *only* fast-forward merge between them. There are two options to do that:

1. `git reset` the `lzma-22.01` branch to the tag (or `whats-next`). This means losing *all* commits that are only on the `lzma-22.01` branch. :warning: Also everyone who has a copy of the branch would have to manually accept the new history, just like after a rebase.
2. Merge `lzma-22.01` into `whats-next` (this could be a ff-merge with the history you showed). Obviously that means your existing merge commits end up in `whats-next` forever.

> What did I do wrong?

You mentioned accepting a fast-forward merge, but there must've been a non-fast-forward merge somewhere. If you're prompted for a commit message when merging, it's not fast-forward.

**Reply by pete4abw** [reply]
Thanks, as always, @airtower-luna . The `whats-next` branch is really a testing area for new features. Somewhere along the line, I must have committed something and then discarded the idea. Merging "a" back into "b" presents no problems since it's a development branch. I do not want to upset the main branch since there are pullers and cloners. When I did the merge, I did `-ff` but not `-ff-only`. Next time I will do a no-commit because I don't care if the branches share identical commits or just make sure the ancestors are the same. Thanks again.
